tests/unit/s2n_send_test.c

```diff
@@ -124,18 +246,360 @@ int main(int argc, char **argv)
+    }
+
+    /* s2n_sendv_with_offset mitigates BEAST with small writes  */
+    {
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+            s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        uint32_t writes = 0;
+
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_send_mitigates_beast_fn));
+        EXPECT_SUCCESS(s2n_connection_set_ctx(conn, (void*)&writes));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
+
+        s2n_blocked_status blocked = 0;
+
+        /* We need to initialize the client mac to make sure our encrypted data is larger than the CBC block size. */
+        struct s2n_hmac_state *mock_mac_state = &conn->client->client_record_mac;
+        EXPECT_SUCCESS(s2n_hmac_init(mock_mac_state, S2N_HMAC_SHA256, test_data, sizeof(test_data)));
+
+        /* Tweak the connection object to enable the BEAST mitigation. 
+         * Currently the conditions we need to hit to enable the BEAST mitigation are:
+         * 1. actual_protocol_version < S2N_TLS11 
+         * 2. connection is in client mode
+         * 3. the chosen cipher uses CBC */
+        conn->actual_protocol_version = S2N_TLS10;
+        /* This can be any cipher suite that uses CBC. s2n_rsa_with_3des_ede_cbc_sha is chosen because it's easier to stub. */
+        struct s2n_cipher_suite *cipher_suite = &s2n_rsa_with_3des_ede_cbc_sha;
+        struct s2n_cipher_suite *composite_cipher_suite = &s2n_rsa_with_aes_128_cbc_sha;
+        conn->client->cipher_suite = cipher_suite;
+
+        /* We need to check out if a composite cipher is available to filter out BoringSSL and AWSLC, or s2n_send will fail trying to use CBC. */
+        if (cipher_suite->available && composite_cipher_suite->all_record_algs[0]->cipher->is_available()) {
+            s2n_custom_send_fn_called = false;
+            EXPECT_SUCCESS(s2n_send(conn, test_data, sizeof(test_data), &blocked));
+            EXPECT_TRUE(s2n_custom_send_fn_called);
+            EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
+        }
+    }
+
+    /* s2n_send partial writes are flushed in proceeding sends */
+    {
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+            s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(conn);
+        EXPECT_OK(s2n_connection_set_secrets(conn));
+
+        uint32_t actual_bytes_written = 0;
+        EXPECT_SUCCESS(s2n_connection_set_ctx(conn, (void*)&actual_bytes_written));
+
+        s2n_blocked_status blocked = 0;
+
+        EXPECT_SUCCESS(s2n_connection_set_send_cb(conn, s2n_partial_socket_send_fn));
+        EXPECT_SUCCESS(s2n_connection_set_send_ctx(conn, (void*) conn));
+        s2n_custom_send_fn_called = false;
+
+        /* Assume we have a partial write of sizeof(test_data) - 1 that was successful.
+         * The subsequent socket write failed with EAGAIN, and control has returned to the application.*/
```

Comment:
Why can we assume this? Doesn't s2n_partial_socket_send_fn write len-3?

Reply:
Ah oops, will update this comment. At one time it was at len-1, and I forgot to update this comment when I changed it